### PR TITLE
Add support for SCD4X

### PR DESCRIPTION
**Pull request by sjtrny**
# What does this implement/fix? 

Adds support for the SCD4X sensor, which is a CO2 sensor (+temp and RH).

There is already a PR for this sensor. However it modifies the SCD30 sensor.

This is a simpler version that provides the same feature set as SCD30 but with the updated SCD40 sensor.

In future better support for temperature/pressure compensation and forced recalibration could be provided.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1429

## Test Environment

- [X] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:

```
  - platform: scd4x
    co2:
      name: "CO2"
    temperature:
      name: "Temperature"
    humidity:
      name: "Humidity"

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by oxan** [comment]
> There is already a PR for this sensor.

That seems to be #1561. 

> However it modifies the SCD30 sensor.
> 
> This is a simpler version that provides the same feature set as SCD30 but with the updated SCD40 sensor.

Could you elaborate a bit on why this PR is needed or better than that other PR? Judging from the code they seem to be 90% similar.

**Comment by sjtrny** [comment]
> > There is already a PR for this sensor.
> 
> That seems to be #1561.
> 
> > However it modifies the SCD30 sensor.
> > This is a simpler version that provides the same feature set as SCD30 but with the updated SCD40 sensor.
> 
> Could you elaborate a bit on why this PR is needed or better than that other PR? Judging from the code they seem to be 90% similar.

Because it modifies an existing sensor definition. I think it's a better idea (hygiene, bug safety) to just have the introduction of a new sensor be standalone. Only if there's enough justification should a merged/shared codebase be used.

I kind of expect uptake of SCD30 sensors to die anyway. Most people will move to the SCD4X series because of size.

File: esphome/components/scd4x/sensor.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                device_class=DEVICE_CLASS_HUMIDITY,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+            cv.Optional(CONF_AUTOMATIC_SELF_CALIBRATION, default=True): cv.boolean,
+            cv.Optional(CONF_ALTITUDE_COMPENSATION, default="0m"): cv.All(
+                cv.float_with_unit("altitude", "(m|m a.s.l.|MAMSL|MASL)"),
+                cv.int_range(min=0, max=0xFFFF, max_included=False),
+            ),
+            cv.Optional(CONF_AMBIENT_PRESSURE_COMPENSATION): cv.pressure,
+            cv.Optional(CONF_TEMPERATURE_OFFSET, default="4°C"): cv.temperature,
```
Comment (OttoWinter) [inline esphome/components/scd4x/sensor.py:64]: This is like a constant offset for the temperature value, right? Does the manufacturer suggest 4°C or is this just a value that worked for you?

Comment (sjtrny) [reply esphome/components/scd4x/sensor.py:64]: > This is like a constant offset for the temperature value, right?

Correct

> Does the manufacturer suggest 4°C or is this just a value that worked for you?

It's the default value for the sensor so yes it's manufacturer recommended. From section 3.6.1 of the [data sheet](https://www.sensirion.com/fileadmin/user_upload/customers/sensirion/Dokumente/9.5_CO2/Sensirion_CO2_Sensors_SCD4x_Datasheet.pdf) (last sentence):

`Note that the temperature offset can depend on various factors such as the SCD4x measurement mode, self-heating of close components, the ambient temperature and air flow. Thus, the SCD4x temperature offset should be determined inside the customer device under its typical operation conditions (including the operation mode to be used in the application) and in thermal equilibrium. Per default, the temperature offset is set to 4° C.`

**Review by OttoWinter — APPROVED** [review]

**Comment by oxan** [comment]
@sjtrny Could you test the component with the changes Otto made? If it still works, this can be merged.

**Comment by sjtrny** [comment]
> @sjtrny Could you test the component with the changes Otto made? If it still works, this can be merged.

Confirmed. It is working as expected 🎉.

```
[19:27:40][D][sensor:121]: 'CO2': Sending state 691.00000 ppm with 0 decimals of accuracy
[19:27:40][D][sensor:121]: 'Temperature': Sending state 22.85736 °C with 2 decimals of accuracy
[19:27:40][D][sensor:121]: 'Humidity': Sending state 51.92871 % with 2 decimals of accuracy
```

**Comment by oxan** [comment]
Thanks!

**Pull request merged**